aws/lambda-with-other-services/send-sms-code-with-sns/send-sms-code-with-sns.md

```diff
@@ -0,0 +1,123 @@
+
+To start, create a lambda function and require the AWS software development kit at the top like so: `const AWS = require('aws-sdk');`. This allows for the use of Amazon Web Services' various platforms. In this case, we will be using AWS.SNS.
+
+The "event" parameter we will be passing into the exports.handler is a JSON object made with a person's information including: first name, last name, phone number, email, etc.
+
+Inside of the exports.handler function, include the SNS event like so: `const SNS = new AWS.SNS();`. This allows the function to use the SNS platform to send and receive SMS messages.
+
+In order to send a text message, we will need to use a Publish action included within the SNS platform's inherent abilities. For JavaScript, the Publish action must take in at least a message and a phone number. Define params to equal an object that holds these keys as shown below:
+
+```javascript
+let params = {
+	'Message': smsCode,
+	'PhoneNumber': event.phoneNumber,
+};
+```
+
+A third key that can be added to the params is the MessageAttributes key, which holds additional information for use by the SNS event.
+- A sender ID can be added to let the SMS show who the message is from. (This does not work within the U.S. For a list of supported countries and regions, please see [here](https://docs.aws.amazon.com/sns/latest/dg/sms_supported-countries.html))
+- A max price can be set so that a message will not send if it would exceed the maximum you are willing to pay for it to succeed. (The price for outbound message now in the U.S. is $0.00645. Prices may vary per provider. To check inbound or other countries and regions prices, please see [here](https://aws.amazon.com/sns/sms-pricing/))
+- The SMS type can be set to Promotional or Transactional. Promotional optimizes for cost efficiency, while Transactional optimizes for reliability.
+
+The syntax for the above is shown below.
+
+```javascript
+let params = {
+    'Message': smsCode,
+    'PhoneNumber': event.phoneNumber,
+    'MessageAttributes': {
+        'AWS.SNS.SMS.SenderID': 'MKDecision'
+        'AWS.SNS.SMS.MaxPrice': 0.00645,
+        'AWS.SNS.SMS.SMSType': 'Transactional'
+    }
+};
+```
+
+These params can now be passed into our Publish action and be promisified which completes the basic set up of our Publish function.
+
+```javascript
+ SNS.publish(params).promise();
+ ```
+
+ This is what we will be returning at the end of our lambda function.
+
+ Next, we can set up a function that will generate a verification code to send to our user once they input their information as shown below.
+
+ ```javascript
+    //...
+    	let smsCode = '';
+
+	//this function will generate a random number with d amount of digits. zeros can be the first digit.
+	function generate(d) {
+
+		let max = Math.pow(10, d)*9-1;
+		let min = Math.pow(10, d);
+
+		let number = Math.floor( Math.random() * max + min);
+		smsCode = ('' + number).substring(1);
```

Comment:
This implementation is okay for the most part, but it would be better if you converted the number directly to a string and made sure that leading 0's are included.